tools/clang/include/clang/SPIRV/Structure.h

```diff
@@ -0,0 +1,305 @@
+  SPIRVModule &operator=(const SPIRVModule &) = delete;
+
+  // Allow move constructor/assignment since they are efficient.
+  SPIRVModule(SPIRVModule &&that) = default;
+  SPIRVModule &operator=(SPIRVModule &&that) = default;
+
+  /// \brief Returns true if this module is empty.
+  bool isEmpty() const;
+  /// \brief Clears all instructions and functions and turns this module into
+  /// an empty module.
+  void clear();
+
+  /// \brief Sets the id bound to the given bound.
+  inline void setBound(uint32_t newBound);
+
+  /// \brief Collects all the SPIR-V words in this module and consumes them
+  /// using the consumer within the given InstBuilder. This method is
+  /// destructive; the module will be consumed and cleared after calling it.
+  void take(InstBuilder *builder);
+
+  inline void addCapability(spv::Capability);
+  inline void addExtension(std::string extension);
+  inline void addExtInstSet(uint32_t setId, std::string extInstSet);
+  inline void setAddressingModel(spv::AddressingModel);
+  inline void setMemoryModel(spv::MemoryModel);
+  inline void addEntryPoint(spv::ExecutionModel, uint32_t targetId,
+                            std::string targetName,
+                            std::initializer_list<uint32_t> intefaces);
+  inline void addExecutionMode(Instruction &&);
+  inline void addDebugName(uint32_t targetId,
+                           llvm::Optional<uint32_t> memberIndex,
+                           std::string name);
+  inline void addDecoration(Instruction &&);
+  inline void addType(Instruction &&);
+  inline void addFunction(Function &&);
+
+private:
+  /// \brief The struct representing a SPIR-V module header.
+  struct Header {
+    /// \brief Default constructs a SPIR-V module header with id bound 0.
+    Header();
+
+    /// \brief Feeds the consumer with all the SPIR-V words for this header.
+    void collect(const WordConsumer &consumer);
+
+    uint32_t magicNumber;
+    uint32_t version;
+    uint32_t generator;
+    uint32_t bound;
+    uint32_t reserved;
+  };
+
+  /// \brief The struct representing an entry point.
+  struct EntryPoint {
+    inline EntryPoint(spv::ExecutionModel, uint32_t id, std::string name,
+                      std::initializer_list<uint32_t> interface);
+
+    spv::ExecutionModel executionModel;
+    uint32_t targetId;
+    std::string targetName;
+    std::initializer_list<uint32_t> interfaces;
+  };
+
+  /// \brief The struct representing a debug name.
+  struct DebugName {
+    inline DebugName(uint32_t id, llvm::Optional<uint32_t> index,
+                     std::string targetName);
+
+    uint32_t targetId;
+    llvm::Optional<uint32_t> memberIndex;
+    std::string name;
+  };
+
+  Header header; ///< SPIR-V module header.
+  std::vector<spv::Capability> capabilities;
+  std::vector<std::string> extensions;
+  std::vector<std::pair<uint32_t, std::string>> extInstSets;
+  llvm::Optional<spv::AddressingModel> addressingModel;
+  llvm::Optional<spv::MemoryModel> memoryModel;
```

Comment:
Yeah, I was thinking about this earlier but still decided to go this way b/c

* It gives a meaningful "zero" value for addressing and memory module, so we can have a meaningful empty module.
* It makes the module a little bit more flexible.

Can definitely change this later.